command/agent/agent_endpoint.go

```diff
@@ -299,3 +299,92 @@ type joinResult struct {
 	NumJoined int    `json:"num_joined"`
 	Error     string `json:"error"`
 }
+
+func (s *HTTPServer) HealthRequest(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	if req.Method != "GET" {
+		return nil, CodedError(405, ErrInvalidMethod)
+	}
+
+	var args structs.GenericRequest
+	if s.parse(resp, req, &args.Region, &args.QueryOptions) {
+		return nil, nil
+	}
+
+	health := healthResponse{}
+	getClient := true
+	getServer := true
+
+	// See if we're checking a specific agent type and default to failing
+	if healthType, ok := req.URL.Query()["type"]; ok {
+		getClient = false
+		getServer = false
+		for _, ht := range healthType {
+			switch ht {
+			case "client":
+				getClient = true
+				health.Client = &healthResponseAgent{
+					Ok:      false,
+					Message: "client not enabled",
+				}
+			case "server":
+				getServer = true
+				health.Server = &healthResponseAgent{
+					Ok:      false,
+					Message: "server not enabled",
+				}
+			}
+		}
+	}
+
+	// If we should check the client and it exists assume it's healthy
+	if client := s.agent.Client(); getClient && client != nil {
+		health.Client = &healthResponseAgent{
+			Ok:      true,
+			Message: "ok",
+		}
+	}
+
+	// If we should check the server and it exists, see if there's a leader
+	if server := s.agent.Server(); getServer && server != nil {
+		health.Server = &healthResponseAgent{
+			Ok:      true,
+			Message: "ok",
+		}
+
+		leader := ""
+		if err := s.agent.RPC("Status.Leader", &args, &leader); err != nil {
+			health.Server.Ok = false
+			health.Server.Message = err.Error()
+		} else if leader == "" {
+			health.Server.Ok = false
+			health.Server.Message = "no leader"
+		}
+	}
+
+	if !health.ok() {
+		// At least one not-ok response, set failing status code
+		resp.WriteHeader(500)
```

Comment:
Use `CodedError`: https://github.com/hashicorp/nomad/blob/4fb2dce22a0c82b969a1715c0f1ac7099cd5e331/command/agent/deployment_endpoint.go#L12

Reply:
To use that and keep the JSON format I'd have to serialize the `healthResponse` struct to a string and use that for an error. This just seemed simpler.

Reply:
Oh but this breaks the Content Type header... fixing